android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java

```diff
@@ -202,18 +251,64 @@ protected void onResume() {
           filename = url.substring(index);
         }
 
+        Intent i;
         Bundle bundle = new Bundle();
-        bundle.putString(KMKeyboardDownloaderActivity.ARG_KEYBOARD,
-          data.getQueryParameter(KMKeyboardDownloaderActivity.KMKey_Keyboard));
-        bundle.putString(KMKeyboardDownloaderActivity.ARG_LANGUAGE,
-          data.getQueryParameter(KMKeyboardDownloaderActivity.KMKey_Language));
-        bundle.putBoolean(KMKeyboardDownloaderActivity.ARG_IS_CUSTOM, isCustom);
-        bundle.putBoolean(KMKeyboardDownloaderActivity.ARG_IS_DIRECT, isDirect);
-        bundle.putString(KMKeyboardDownloaderActivity.ARG_URL, url);
-        bundle.putString(KMKeyboardDownloaderActivity.ARG_FILENAME, filename);
-        Intent i = new Intent(getApplicationContext(), KMKeyboardDownloaderActivity.class);
-        i.putExtras(bundle);
-        startActivity(i);
+        if (data.getScheme().equals("file") && kmpFile.exists()) {
+          // KMP is already local. Copy KMP to app cache and start PackageActivity
+          File cacheKmpFile = new File(MainActivity.this.getCacheDir().toString(), kmpFile.getName());
+          if (cacheKmpFile.exists()) {
+            cacheKmpFile.delete();
+          }
+          try {
+            FileUtils.copy(kmpFile, cacheKmpFile);
+          } catch (Exception e) {
+            Log.e("onResume", "Copy failed: " + e);
+            intent.setData(null);
+            return;
+          }
+
+          bundle.putString("kmpFile", cacheKmpFile.getAbsolutePath());
+          Intent packageIntent = new Intent(getApplicationContext(), PackageActivity.class);
+          packageIntent.putExtras(bundle);
+          startActivity(packageIntent);
+        } else if (url.endsWith(".kmp")) {
+          try {
+            // Download the KMP to app cache
```

Comment:
Will this work with `file:` files that don't exist?

Reply:
That scenario would go down this `else if` branch and fail.

I'm addressing this in the "peer-to-peer KMP distribution" work, where a switch statement separates the protocols and `file://` only handles `.kmp` files